extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/TransformedAnnotationsBuildItem.java

```diff
@@ -48,7 +57,39 @@ public boolean hasAnnotation(AnnotationTarget target, Class<? extends Annotation
 
     @Override
     public Collection<AnnotationInstance> apply(AnnotationTarget target) {
-        return beanDeployment.getAnnotations(target);
+        return queryAndConditionallyFilter(target);
     }
 
+    private Collection<AnnotationInstance> queryAndConditionallyFilter(AnnotationTarget target) {
+        if (target.kind().equals(AnnotationTarget.Kind.METHOD_PARAMETER)) {
+            // We cannot query Jandex for method param. annotation target, so we operate on the whole method
+            // and filter results accordingly
+            Collection<AnnotationInstance> result = new ArrayList<>();
+            for (AnnotationInstance instance : beanDeployment.getAnnotations(target.asMethodParameter().method())) {
+                if (instance.target().kind().equals(AnnotationTarget.Kind.METHOD_PARAMETER)
+                        && instance.target().asMethodParameter().position() == target.asMethodParameter().position()) {
+                    result.add(instance);
+                }
+            }
+            return result;
+        } else {
+            return beanDeployment.getAnnotations(target);
+        }
+    }
+
+    private AnnotationInstance queryForMethodParam(AnnotationTarget target, DotName annotationName) {
+        if (!target.kind().equals(AnnotationTarget.Kind.METHOD_PARAMETER)) {
+            throw new IllegalArgumentException(
+                    "TransformedAnnotationsBuildItem#queryForMethodParam needs to operate on METHOD_PARAMETER AnnotationTarget");
+        }
+        // We cannot query Jandex for method param. annotation target, so we operate on the whole method
+        // and filter results accordingly
+        AnnotationInstance instance = beanDeployment.getAnnotation(target.asMethodParameter().method(), annotationName);
```

Comment:
This doesn't necessarily have to work. First, the method can return `null`, which the present code doesn't account for, and second, the method can return an annotation present on a different target than expected, which doesn't tell you anything about presence or absence of an annotation on the expected target.

What should work is first doing `beanDeployment.getAnnotations()` and then filtering on the result.

Reply:
Good point, done